core/commands/up_command.rb

```diff
@@ -226,22 +246,26 @@ def check_and_configure_nodes(nodes_to_check)
     halt_nodes.concat(configure_nodes(unconfigured_nodes))
   end
 
-  # Destroy all existing nodes and setup configuration
+  # Starts and configurats mathing nodes from configuration file
   #
   # @param config [Configuration] configuration that should be run
   # @param node [String] name of the node to bring up
-  # @return [Array<String>] list of nodes that should be checked
+  # @return [Array<String>] list of nodes that should be fixed
   def setup_nodes(config, node = '')
     generate_docker_images(config.template, '.') if config.provider == 'docker'
-    @ui.info 'Destroying existing nodes.'
-    run_command_and_log("vagrant destroy --force #{node}")
-    bring_up_machines(config.provider)
-    nodes_to_check = if node.empty?
-                       config.node_names
-                     else
+    nodes_to_check = if !node.empty?
```

Comment:
Did you consider to move whole this logic to the the Configuration class? Therefore it would parse the template string, also the --labels parameter. This way we could easily re-use this logic in other methods that deal with the configuration as a whole.